### ISSUE TITLE
Memory safety: addresses of free'd objects are excluded from the address space

### ISSUE THREAD

**Issue by fbrausse** [esbmc/esbmc#775]
This affects the memsafety benchmark [c/memsafety/cmp-freed-ptr](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/memsafety/cmp-freed-ptr.yml) ([C source](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/memsafety/cmp-freed-ptr.c)) which ESBMC wrongly recognizes as `VERIFICATION SUCCESSFUL`:

```c
#include <stdlib.h>
#include <stdint.h>

int main() {

    struct T {
        struct T* next;
    };

    struct T* x = NULL;
    struct T* y = NULL;

    y = malloc(sizeof(*y));
    intptr_t adressY = (intptr_t) y;

    free(y);

    x = malloc(sizeof(*x));
    intptr_t adressX = (intptr_t) x;

    if (adressX == adressY)
    { // if the second malloc returns the same value as the first, I should get here
        free(x);
    }

    free(x);

    return 0;
}

// predator-regre/test-0238.c
// - comparing freed pointers
// - contributed by Ondra Lengal
```

> `esbmc --no-div-by-zero-check --force-malloc-success --state-hashing --add-symex-value-sets --k-step 2 --floatbv --unlimited-k-steps sv-benchmarks/c/memsafety/cmp-freed-ptr.c --32  --memory-leak-check`

The formula generated contains
```smt2
(assert (or (bvult __ESBMC_ptr_obj_end_3 __ESBMC_ptr_obj_start_2)
    (bvugt __ESBMC_ptr_obj_start_3 __ESBMC_ptr_obj_end_2)))
```
asserting that the (free'd) object `*y` and the existing object `*x` do not overlap. A guarantee not given by `malloc()`. Indeed, removing this assertion from the formula makes it SAT and ESBMC thus correctly finding the memsafety violation.

This assertion is generated in `smt_convt::finalize_pointer_chain()` unconditionally for all previous objects. If it was guarded by a condition saying "object is still alive", I believe this would fix the issue.

**Comment by lucasccordeiro** [comment]
>  If it was guarded by a condition saying "object is still alive", I believe this would fix the issue.

I agree.

**Comment by fbrausse** [comment]
What would be the best way to get the **current** `__ESBMC_alloc` symbol into `smt_convt::finalize_pointer_chain()`? As I understand it, the symbol's name is stored in https://github.com/esbmc/esbmc/blob/e2113d2a54acc456ebc801ede3c9c05341ae33f1/src/goto-symex/goto_symex.h#L816 but `smt_convt` does not have access to it, nor to its current name in the symbolic execution at the point of allocation.

**Comment by lucasccordeiro** [comment]
Hi @fbrausse,

What do you think if we produce the equation at the symex level? I believe our SMT backend should only convert the equations in SSA to SMT. This `finalize_pointer_chain()` sounds a bit weird to me :-)

**Comment by fbrausse** [comment]
`smt_convt::convert_typecast_to_ptr()` might be another related problem if we allow re-allocation of another object at overlapping addresses. This function should actually be part of smt_memspace.cpp and it assumes that all objects (as numbered by smt_memspace.cpp) except INVALID are uniquely identified by their address range in order to re-create the pointer target. It produces an if-then-else chain very similar to what dereference() is doing, though dereference() typically uses `same_object2t` instead of the addresses. I'm not sure whether restricting this chain to only alive objects is sound.

> What do you think if we produce the equation at the symex level? I believe our SMT backend should only convert the equations in SSA to SMT. This finalize_pointer_chain() sounds a bit weird to me :-)

@lucasccordeiro It would definitely be a better fit in symex, yes. We would need to include all live objects in this assertion. smt_memspace.cpp can do it since it knows the total number of object IDs allocated by it so far - making the module itself dependent on the symex sequence as it is now. I think it would be good to decouple that.

**Comment by fbrausse** [comment]
Another unwanted side effect of the current "steal address space" approach:
```c
#include <stdlib.h>

void steal_addr_space(size_t n)
{
	char *a = malloc(n);
	free(a);
}

int main()
{
	size_t MiB = 1ULL << 20;
	/* 16 * 256 MiB = 4 GiB */
	for (int i=0; i<16; i++)
		steal_addr_space(256 * MiB);
	assert(0);
}
```
is `VERIFICATION SUCCESSFUL` (after about 12 minutes) when run with `--32 --force-malloc-success`.

**Comment by fbrausse** [comment]
>> This assertion is generated in smt_convt::finalize_pointer_chain() unconditionally for all previous objects. If it was guarded by a condition saying "object is still alive", I believe this would fix the issue.

PRs #781 and #788 tried this approach, and while both can be made to work for dynamic allocations (the SMT one for instance by modifying the slicing), I now believe this approach is only part of a proper solution:

The smt-memspace does not only handle dynamic allocations, but all kinds of objects whose addresses have been taken. This can be explicit via addr-of or implicit by malloc/realloc. All those symbols (as well as NULL and INVALID) get assigned "pointer object" numbers and the alive ones of them all should have unique addresses.

> I believe our SMT backend should only convert the equations in SSA to SMT. This finalize_pointer_chain() sounds a bit weird to me :-)

These objects are passed to smt-conv in the sequence the SSA describes and smt-memspace assumes that all previous objects are alive when a new one is observed in an expression. If smt-conv did "only convert the equations in SSA to SMT" it  would have no way of knowing which previous objects are dead or free'd, and this includes both, dynamically allocated ones and symbols local to a closed frame. This information is just not communicated to it.

In particular, this is `VERIFICATION SUCCESSFUL` since `*p` is assumed alive when `q` is assigned - they can't have the same address:
```c
int * f()
{
	int x[1];
	return x;
}

int main()
{
	int *p = f();
	int *q = f();
	assert(p != q);
}
```

As I see it, we either have the option (a) to be explicit about the dead/free'd objects in the communication to smt-conv (for instance by passing such an object->alive mapping to convert_ast() or by giving a callback to the smt_convt constructor), or (b) to model the memspace entirely on symex level, which I think would require us to get rid of any expressions involving (implicitly or explicitly taken) object addresses. Option (b) would also need to ensure that all threads operate on the same single heap.

Did I miss something? What do you think of these options?

In https://github.com/esbmc/esbmc/pull/788#pullrequestreview-1029732715 @lucasccordeiro wrote
> We could investigate the two ideas:
> 1. Use an "address set analysis" to obtain an over-approximation of the address space layout constraints. For example, we could look at how we perform value set analysis in the pointer-analysis module.
> 2. Use a lightweight incremental SMT solving or the UNSAT core produced by the SMT solver to obtain information about the existing dynamic objects.

Both cases would still need a choice between (a) and (b) and case 2. should be extended to cover not only dynamic objects, see the example above.

**Comment by lucasccordeiro** [comment]
@fbrausse: can I ask you for the status of this issue?

**Comment by lucasccordeiro** [comment]
@fbrausse: can I ask you for the status of this issue?

**Comment by fbrausse** [comment]
@lucasccordeiro There are two PRs attempting to fix this issue: #781 (smt layer) and #785 (symex). The first I closed as it was deemed "too hacky", and the second had a too large performance impact. A possibility to combine both approaches could be to introduce a new statement during symex which gets translated into the required assertions about alive objects not overlapping.

The "too hacky" part above is a side effect of object's pointer regions being encoded purely by smt_memspace. As long as that's done there I'm very open to other suggestions how to approach this problem.

From https://github.com/esbmc/esbmc/issues/775#issuecomment-1179712955:
> As I see it, we either have the option (a) to be explicit about the dead/free'd objects in the communication to smt-conv (for instance by passing such an object->alive mapping to convert_ast() or by giving a callback to the smt_convt constructor), or (b) to model the memspace entirely on symex level, which I think would require us to get rid of any expressions involving (implicitly or explicitly taken) object addresses. Option (b) would also need to ensure that all threads operate on the same single heap.

I'm currently tending towards option (a) and, secondly, also refining the mental model that the smt-conv layer only processes equations - atm. it does not, it tracks pointed-to objects in the SSA sequence. With option (a) that would be more explicit, e.g., if there was a separate "kill object" statement given to it whenever an object leaves the scope, is free'd, or similar.

**Comment by lucasccordeiro** [comment]
@fbrausse: let's move forward with option (a).
